Add script for verification of junit xml reports

PerMac opened the pull request:
The script is called by:
python3 results_verification.py -P <PATH_TO_REPORT VERSION> -Z <ZEPHYR_VERSION>

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

maksimmasalski commented:
@hakehuang

abrodkin commented:
Would be good to add @wangnuannuan to reviewers.

PerMac commented:
> Would be good to add @wangnuannuan to reviewers.

Not sure how this repo is set up. I cannot do this. I can add only some of you... (eg. I cannot add Hake, Alexey)

erwango commented on scripts/results_verification.py:
```diff
@@ -0,0 +1,74 @@
+def main(path, version):
+    # TODO: add argparse
+    max_size = 5
+    max_failures = 50
+    max_errors = 50
```
It seems reasonnable to add such limits.
Though, I'm affraid this could be a bit low when you have multiple boards on a bench.
I should have at least 15 boards on the bench (in good days).
If there are 4 tests failing on all these boards as this could happen for a kernel test that may impact all the boards in the same way I'm above this limit...
Or maybe there is a a way not to count multiple failures for the same test?

PerMac replied:
Right now I see a workflow in a way that this script will be called for every file in a commit (one file=one platform). So it is actually a limit per platform. And of course, the numbers are now just from the top of my head. I will add argparsing so the numbers can be directly addressed when the script is called.

erwango replied:
Ok, sure if that's per boards it's reasonable!

PerMac commented on scripts/results_verification.py:
```diff
@@ -0,0 +1,74 @@
+
+    file_path = Path(path)
+
+    if not os.path.exists(file_path):
```
I think your idea won't work with the publishing workflow. In a single folder (corresponding to a given zephyr version) there will be reports from all platforms, as one adds their results to this folder. The script would then run on the same files multiple times with each commit targeting the same version. IMO it is cleaner if some higher layer decides which file to verify and this could be done in github actions by scanning the xml files which are updated in the PR.

hakehuang replied:
GHAction check file one by one, but if I want to try it on my local first, I need run for multiply times.

PerMac commented on scripts/results_verification.py:
```diff
@@ -0,0 +1,74 @@
+    if not os.path.exists(file_path):
+        print(f"XML report not found at {file_path}")
+        raise FileNotFoundError
+    assert check_version(file_path, version), "Incorrect version"
```
I'm sorry, I don't understand what do you mean. Can you explain it in other words?

hakehuang replied:
this is a design pattern, see https://refactoring.guru/design-patterns/proxy/python/example. maybe looks too complex for now. just query whether you have ready solution.

PerMac commented on scripts/results_verification.py:
```diff
@@ -0,0 +1,74 @@
+
+if __name__ == "__main__":
+    main(path=sys.argv[1], version=sys.argv[2])
+
```
Should be fine now

PerMac reviewed (commented):
@hakehuang Why did you change the version verifying function? IMO the original workflow is cleaner. This script doesn't need the steps for reading the version from web. You tell it which version should the checked reports have, and it does exactly that.
In the current publishing workflow the version in the report has to be the same as the folder name, which a commit with results is targeting. IMO it is the cleanest if this part is handled by github actions, as all information should already be available on that level. This way the publishing is more modular. E.g. recently the format of versions.json has changed. With your solution, one will also have to change the results verification script to adapt to such change.

PerMac commented:
@hakehuang your change makes the script only checking if the version in the report exists in the list of daily versions. My solution allows verifying if the PR is targeting the correct version of the results.

PerMac commented:
I move the PR out from the draft and reverted Hake's commit, @nashif Can you review it?

nashif commented:
Would be nice to see this integrated in the action as well, for example:

```
diff --git a/.github/workflows/check_results.yml b/.github/workflows/check_results.yml
index 44b4cff..bea2adf 100644
--- a/.github/workflows/check_results.yml
+++ b/.github/workflows/check_results.yml
@@ -48,6 +48,12 @@ jobs:
             echo "File ${file} not allowed."
             exit 1
           fi
+          if [[ ${file} == results/* ]]; then
+            echo "Verifying content of ${file}"
+            fd=$(dirname $file)
+            v=$(basename $fd)
+            python3 scripts/results_verification.py -V ${v} -P ./${file}
+          fi
           if [[ ${file} == results/* ]]; then
             echo "Checking ${file}"
             ~/.local/bin/junit2html --summary-matrix ${file}

```

nashif commented on scripts/results_verification.py:
```diff
@@ -0,0 +1,98 @@
+    """Parse and return required limits from arguments"""
+    argpar = ArgumentParser(description='Verifies if a given report fulfils requirements before its publishing')
+    argpar.add_argument('-P', '--path', required=True, type=str, help='Path to the report which will be verified')
+    argpar.add_argument('-V', '--version', required=True, type=str, help='Version to be verified')
```
--version is usually used to print out the version of the script itself, pick something else please

PerMac replied:
Changed to '-Z'/ '--zephyr'

hakehuang commented:
> @hakehuang Why did you change the version verifying function? IMO the original workflow is cleaner. This script doesn't need the steps for reading the version from web. You tell it which version should the checked reports have, and it does exactly that.
> In the current publishing workflow the version in the report has to be the same as the folder name, which a commit with results is targeting. IMO it is the cleanest if this part is handled by github actions, as all information should already be available on that level. This way the publishing is more modular. E.g. recently the format of versions.json has changed. With your solution, one will also have to change the results verification script to adapt to such change.

There is a possibility that we will have version that can not found in the https://testing.zephyrproject.org/daily_tests/index.html. if this can be accept then we do not need to check the version.json

PerMac commented:
> There is a possibility that we will have version that can not found in the https://testing.zephyrproject.org/daily_tests/index.html. if this can be accept then we do not need to check the version.json

Ok, I added this check as well.

PerMac commented:
I modified the check_results.yml according to your suggestion, just added `exit 1` in case of verification failure. I also added a file with 100 errors and you can see the check failed as it should for this PR. @nashif Let me know if you accept the changes and I will remove these test XML files from the PR.

nashif commented:
@PerMac  you are adding some results here that make it fail, is this intended, otherwise looks good.

PerMac commented:
@nashif, yes it was intentional to test/show the rejection of a file. I removed it now so PR should be ready to merge
